ReactiveCocoa/DelegateProxy.swift

```diff
@@ -1,106 +1,259 @@
 import ReactiveSwift
 import enum Result.NoError
 
-internal class DelegateProxy<Delegate: NSObjectProtocol>: NSObject {
-	internal weak var forwardee: Delegate? {
-		didSet {
-			originalSetter(self)
-		}
+private let delegateProxySetupKey = AssociationKey<Bool>(default: false)
+private let hasSwizzledKey = AssociationKey<Bool>(default: false)
+
+internal protocol DelegateProxyProtocol: class {
+	var _forwardee: AnyObject? { get }
+	func runtimeWillIntercept(_ selector: Selector, signal: Signal<AnyObject, NoError>) -> Signal<AnyObject, NoError>
+}
+
+public struct DelegateProxyConfiguration {
+	fileprivate let objcProtocol: Protocol
+	fileprivate let lifetime: Lifetime
+	fileprivate let originalSetter: (AnyObject) -> Void
+}
+
+public class DelegateProxy<Delegate: NSObjectProtocol>: NSObject, DelegateProxyProtocol {
+	public final var delegateType: Delegate.Type {
+		return Delegate.self
+	}
+
+	internal final weak var _forwardee: AnyObject?
+
+	public weak var forwardee: Delegate? {
+		get { return _forwardee as! Delegate? }
+		set { _forwardee = newValue; originalSetter(self) }
 	}
 
-	internal var interceptedSelectors: Set<Selector> = []
+	internal final let lifetime: Lifetime
+
+	private final let writeLock = NSLock()
+	private final var interceptedSelectors: Set<Selector>
+	private final let originalSetter: (AnyObject) -> Void
+	private final let objcProtocol: Protocol
+
+	public required init(config: DelegateProxyConfiguration) {
```

Comment:
Should this be `public`? Or should we only expose the `proxy` method below?

If this is `public`, should we inline the properties from `DelegateProxyConfiguration` in the `init`?

Reply:
The initialiser is meant to be exposed, while the details of the configuration are meant to be hidden. This is to satisfy the compiler constraint of instantiating instances from the metatype of the non-final `DelegateProxy`, which requires `init(DelegateProxyConfiguration)` to be always present in all subclasses, so that the instantiation via the metatype always succeeds.

Reply:
Why is `DelegateProxy` non-final?